File: platform-common/src/main/java/org/sweetchips/platform/common/Transformer.java
package org.sweetchips.platform.common;

import org.sweetchips.utility.AsyncUtil;
import org.sweetchips.utility.StageWorker;

import java.util.Collection;

final class Transformer {

    private static final String TAG = "Transformer";

    private final ContextLogger mLogger;
    private final StageWorker mBarrier;
    private final Collection<RootUnit> mUnits;

    public Transformer(ContextLogger logger, StageWorker barrier, Collection<RootUnit> units) {
        mLogger = logger;
        mBarrier = barrier;
        mUnits = units;
    }

    public void doWork() {
        mLogger.d(TAG, "doWork: begin");
        AsyncUtil.managedBlock(() -> {
            mBarrier.await();
            doPrepare();
            mBarrier.await();
            doTransform();
            mBarrier.await();
        });
        mLogger.d(TAG, "doWork: end");
    }

    private void doPrepare() {
        mLogger.d(TAG, "doPrepare: begin");
        AsyncUtil.with(mUnits.stream())
                .forkJoin(it -> {
                    switch (it.getStatus()) {
                        case REMOVED:
                        case NOTCHANGED:
                            return;
                        case ADDED:
                        case CHANGED:
                            it.onPrepare();
                            return;
                        default:
                            throw new IllegalArgumentException(it.toString());
                    }
                });
        mLogger.d(TAG, "doPrepare: end");
    }

    private void doTransform() {
        mLogger.d(TAG, "doTransform: begin");
        AsyncUtil.with(mUnits.stream())
                .forkJoin(it -> {
                    switch (it.getStatus()) {
                        case REMOVED:
                            it.onDelete();
                            return;
                        case NOTCHANGED:
                            return;
                        case ADDED:
                        case CHANGED:
                            it.onTransform();
                            return;
                        default:
                            throw new IllegalArgumentException(it.toString());
                    }
                });
        mLogger.d(TAG, "doTransform: end");
    }
}

File: platform-common/src/main/kotlin/org/sweetchips/platform/common/Transformer.kt
package org.sweetchips.platform.common

import kotlinx.coroutines.*
import org.sweetchips.utility.FilesUtil
import java.lang.Runnable

internal class Transformer(
    private val mUnits: Collection<RootUnit>,
) {

    companion object {
        private const val TAG = "Transformer"

        internal fun Collection<Transformer>.doWork(
            prepareBefore: Runnable,
            prepareAfter: Runnable,
            transformBefore: Runnable,
            transformAfter: Runnable,
            logger: ContextLogger,
        ) {
            logger.d(TAG, "doWork: begin")
            prepareBefore.run()
            work(logger, "doPrepare") { doPrepare() }
            prepareAfter.run()
            transformBefore.run()
            work(logger, "doTransform") { doTransform() }
            transformAfter.run()
            logger.d(TAG, "doWork: end")
        }

        private fun Collection<Transformer>.work(
            logger: ContextLogger,
            tag: String,
            work: suspend Transformer.() -> Unit
        ) {
            logger.d(TAG, "$tag: begin")
            runBlocking {
                forEach {
                    launch {
                        it.work()
                    }
                }
            }
            logger.d(TAG, "$tag: end")
        }
    }

    private suspend fun doPrepare() = supervisorScope {
        mUnits.forEach {
            when (it.status) {
                RootUnit.Status.REMOVED, RootUnit.Status.NOTCHANGED -> Unit
                RootUnit.Status.ADDED, RootUnit.Status.CHANGED -> launch {
                    it.onPrepare()
                }
            }
        }
    }

    private suspend fun doTransform() = supervisorScope {
        mUnits.forEach {
            when (it.status) {
                RootUnit.Status.REMOVED -> launch(Dispatchers.IO) {
                    FilesUtil.deleteIfExists(it.output)
                }
                RootUnit.Status.NOTCHANGED -> Unit
                RootUnit.Status.ADDED, RootUnit.Status.CHANGED -> launch {
                    it.onTransform()
                }
            }
        }
    }
}